santoku/utils/configuration.py

```diff
@@ -0,0 +1,347 @@
+        True.
+
+        The given `configuration` must follow the given schema, if any.
+
+        Parameters
+        ----------
+        name : str
+            ID of the configuration.
+        settings : Union[dict, list]
+            An arbitrary JSON object, parsed to use dictionaries and lists.
+        override : bool, optional
+            If set to True, previous configurations with the same name will be overridden.
+
+        Raises
+        ------
+        ConfigurationAlreadyDefined
+            If `override` is set to False and a configuration with the same name already exists.
+
+        InvalidConfiguration
+            If the configuration does not follow the schema.
+        """
+        if not override and name in self.configurations:
+            raise ConfigurationAlreadyDefined(name=name)
+
+        self.validate_schema(settings=settings)
+        self.configurations[name] = settings
+
+        # If the modified config is the active one, reapply it
+        if name == self.active_configuration:
+            self.apply_configuration(name=name)
+
+    def get_configuration(self, name: str) -> JSON:
+        """
+        Retrieves all the settings of an available configuration.
+
+        Parameters
+        ----------
+        name : str
+            Name of the configuration.
+
+        Returns
+        -------
+        Union[dict, list]
+            Arbitrary JSON-like object with the contents of the requested configuration.
+
+        Raises
+        ------
+        UndefinedConfiguration
+            If no configuration with that name exists.
+        """
+        try:
+            return self.configurations[name]
+        except KeyError:
+            raise UndefinedConfiguration(name=name)
+
+    def apply_configuration(self, name: str) -> None:
+        """
+        Sets a given configuration as the active configuration.
+
+        Parameters
+        ----------
+        name: str
+            ID of the configuration.
+
+        Raises
+        ------
+        UndefinedConfiguration
+            If no configuration with that name exists.
+        """
+        if name not in self.configurations:
+            raise UndefinedConfiguration(name=name)
+        self.active_configuration = name
+
+    def get_active_configuration(self) -> JSON:
+        """
+        Retrieve all settings from the active configuration.
+
+        Returns
+        -------
+        Union[dict, list]
+            Arbitrary JSON-like object with the contents of the active configuration.
+
+        Raises
+        ------
+        NoActiveConfiguration
+            If no configuration is set to active
+        UndefinedConfiguration
+            If the active configuration is not available.
+        """
+        if not self.active_configuration:
+            raise NoActiveConfiguration
+        return self.get_configuration(name=self.active_configuration)
+
+    def get_setting(self, key: Any) -> Any:
```

Comment:
Let's imagine this case:
{
  "a": 1.
  "b": {
    "c": 2,
    "d": 3
  }
}

Can I query b and thus retrieving the dictionary containing c and d?

Reply:
Sure! It's already supported.

Reply:
Well, we thought of ways to be able to "lazily" serve values that are not in a leaf node, but all of them were (very) terrible. We were convinced that trying to retrieve many values deep in the nesting if we didn't allow you to get intermediate values would be very tedious, so we ended up letting the user get intermediate results.

Now... I just realized that you can do something like:

```JSON
{
    "a": {
        "b": {
            "c": {
                "d": {
                    "e": 1,
                    "f": 1,
                    "g": 1
                }
            }
        }
    }
}
```

```python
keys_to_get_to_penultimate_level = ["a", "b", "c", "d"]
leaf_node_keys = ["e", "f", "g"]
deeply_nested_settings = {}
for key in leaf_node_keys:
    deeply_nested_settings[key] = config_manager.get_setting(keys_to_get_to_penultimate_level + [key])
```

so it is not too bad after all for a worse case scenario. Thus, I now think it makes sense to except if the user wants to access intermediate levels of the hierarchy.